### ISSUE TITLE
Deadlock: Concat(N) + SelectMany(IO->IE) + SelectMany(IO->Task) + CompletedSynchronously

### ISSUE THREAD

**Issue by RxDave** [dotnet/reactive#318]
Hey everyone,

I just found a deadlock in my application originating from a race condition in a fairly simple Rx query. (v3.1.1, net461)

Is this a known issue?

Here's an example:

```csharp
(from _ in nextInnerQuery
 select from x in Observable.Return(Unit.Default)
        from y in GetEnumerable(x)  // Func<T, IEnumerable<U>>
        from z in MayCompleteSynchronouslyOrAsync(y)  // Func<T, Task>
        select z)
.Concat()
.Subscribe(Console.WriteLine);
```

What seems to be happening:

- The N-ary `Concat` operator acquires a lock while subscribing to an inner observable. (Technically, it's the `Merge(maxConcurrent: 1)` operator acquiring a lock).
- The N-ary `Concat` operator also acquires this same lock when pushing values to its observer's `OnNext` method.
- The `Task`-projected `SelectMany` operator overload acquires a lock when calling `OnNext`.

The order in which these locks are acquired is reversed depending upon whether the projected `Task` completes synchronously or not.

For example: 

- Upon subscribing to my inner query, `Concat`'s lock is acquired and the second `SelectMany` synchronously begins iterating the collection.
- The first value is projected into a `Task` that completes _asynchronously_.
- The second value is projected into a `Task` that completes _synchronously_.
- There is a race condition as to which `Task` will acquire `SelectMany`'s lock first when calling `OnNext`.
- If the second `Task`, the synchronous one, acquires it first, then it pushes the value without a problem 
as it already holds both locks. In this case, there's also no issue for the first `Task` to complete. No deadlock.
- Alternatively, if the first `Task`, the asynchronous one, acquires it first, then it blocks while trying to obtain `Concat`'s lock since it's required by the call to `OnNext`. Now this thread is holding `SelectMany` but waiting to acquire `Concat`.
- Then, the second `Task` attempts to acquire `SelectMany` and blocks. Now this thread is holding `Concat` and waiting to acquire `SelectMany`.
- Deadlock.

Here's a repro:

```csharp
using System;
using System.Collections.Generic;
using System.Reactive;
using System.Reactive.Linq;
using System.Reactive.Subjects;
using System.Threading;
using System.Threading.Tasks;

namespace Labs.Rx
{
  internal static class DeadlockReproLab
  {
    public static void Main()
    {
      CountdownEvent printed = null;
      var nextInnerQuery = new Subject<Unit>();
      IEnumerable<object> e = null;
      Action<string> afterSelectManyButBeforeConcat = _ => { };

      (from _ in nextInnerQuery
       select (from __ in Observable.Return(Unit.Default)
               from completeSynchronously in e
               from x in SyncOrAsync(completeSynchronously as string, completeSynchronously as IObservable<string>)
               select x)
               .Do(x => afterSelectManyButBeforeConcat(x)))
      .Concat()
      .Subscribe(x =>
      {
        Console.WriteLine(x);
        printed.Signal();
      });


      Console.WriteLine("No deadlock...");

      printed = new CountdownEvent(2);
      e = new[] { "Sync 1", "Sync 2" };
      nextInnerQuery.OnNext(Unit.Default);

      printed.Wait();


      Console.WriteLine("Also no deadlock...");

      printed = new CountdownEvent(2);
      var first = new Subject<string>();
      var second = new Subject<string>();
      e = new[] { first, second };
      nextInnerQuery.OnNext(Unit.Default);
      first.OnNext("Async 1");
      second.OnNext("Async 2");

      printed.Wait();


      Console.WriteLine("DEADLOCK");

      printed = new CountdownEvent(2);
      var waitForAsyncOnThreadB = new ManualResetEventSlim();
      var asyncValue = new Subject<string>();

      // Thread B
      afterSelectManyButBeforeConcat = _ => waitForAsyncOnThreadB.Set();

      e = AsyncThenSync(
        asyncValue,
        "Sync 2",
        () =>
        {
          // Thread A
          asyncValue.OnNext("Async 1");
          // Thread B: SelectMany.OnNext acquires a lock
          //           (waitForAsyncOnThreadB is set)
          //           Concat blocks awaiting the lock acquired initially by Thread A.
          //           Order of locks: SelectMany -> Concat (DEADLOCK)

          // Thread A
          waitForAsyncOnThreadB.Wait();
        });

      // Thread A
      nextInnerQuery.OnNext(Unit.Default);   // Thread A: Concat (technically, Merge(1)) acquires and holds a lock while iterating 'e'

      // This line is never reached, and nothing is ever printed.
      printed.Wait();
    }

    private static IEnumerable<object> AsyncThenSync(IObservable<string> asyncValue, string syncValue, Action waitForAsyncOnThreadB)
    {
      // Thread A
      yield return asyncValue;

      waitForAsyncOnThreadB();

      // Thread A
      yield return syncValue;
      // Thread A: SelectMany.OnNext blocks awaiting the lock acquired previously by Thread B.
      //           Order of locks: Concat -> SelectMany (DEADLOCK)
    }

    private async Task<string> SyncOrAsync(string completeSynchronously, IObservable<string> onCompleted)
         => completeSynchronously ?? await onCompleted.Take(1);
  }
}
```

**Comment by RxDave** [comment]
### Workaround: 
It looks like adding `.ToObservable(Scheduler.Default)` to the enumerable avoids the deadlock, for now.

```csharp
(from _ in nextInnerQuery
 select from __ in Observable.Return(Unit.Default)
        from x in GetEnumerable(x).ToObservable(Scheduler.Default)  // Func<T, IEnumerable<U>>
        from y in MayCompleteSynchronouslyOrAsync(x)  // Func<T, Task>
        select y)
.Concat()
.Subscribe(Console.WriteLine);
```

**Comment by danielcweber** [comment]
@RxDave So was it real bug or rather one of those cases where a subscription can't be cancelled because it's emitting items synchronously?

**Comment by RxDave** [comment]
I'm not totally sure, but I suspect that it's a bug in `Concat`. Perhaps it's worth taking a look at why `Concat` locks on the same object when calling `OnNext` and when subscribing to observables?

**Comment by akarnokd** [comment]
I'd say so. Switching from `Concat` to [this `Concat`](https://github.com/dotnet/reactive/pull/491) does not deadlock.

**Comment by akarnokd** [comment]
Closing as most likely resolved by #491.
